File: java/src/main/java/problems101_200/LruCache.java
package problems101_200;

import java.util.LinkedHashMap;
import java.util.Map;

class LruCache {
    int capacity;
    Map<Integer, Integer> cache = new LinkedHashMap<>();

    LruCache(int capacity) {
        this.capacity = capacity;
    }

    int get(int key) {
        if (!cache.containsKey(key)) {
            return -1;
        }
        int value = cache.remove(key);
        cache.put(key, value); // put on end
        return value;
    }

    void put(int key, int value) {
        if (cache.containsKey(key)) {
            cache.remove(key);
        } else if (cache.size() == capacity) {
            cache.remove(cache.keySet().iterator().next());
        }
        cache.put(key, value);
    }
}

File: kotlin/src/main/kotlin/problems101_200/LruCache.kt
package problems101_200

class LruCache(val capacity: Int) {
    val cache = linkedMapOf<Int, Int>()

    operator fun get(key: Int): Int {
        if (!cache.containsKey(key)) {
            return -1
        }
        val value = cache.remove(key)!!
        cache[key] = value // put on end
        return value
    }

    operator fun set(key: Int, value: Int) {
        if (cache.containsKey(key)) {
            cache -= key
        } else if (cache.size == capacity) {
            cache -= cache.keys.iterator().next()
        }
        cache[key] = value
    }
}
